test/compiler/irpasses.jl

```diff
@@ -1881,3 +1881,32 @@ end
 @test !fully_eliminated() do
     all(iszero, Iterators.repeated(0))
 end
+
+## Test that domsort doesn't mangle single-argument phis (#29262)
+let code = Any[
+        # block 1
+        GotoIfNot(Core.Argument(2), 4),
+        # block 2
+        Expr(:call, Base.throw, "error"), # an implicit `unreachable` terminator
+        # block 3
+        Expr(:call, :opaque),
+        # block 4
+        ReturnNode(nothing),
+    ]
+    ir = make_ircode(code; ssavaluetypes=Any[Union{}, Union{}, Any, Union{}])
+
+    # Unfortunately `compute_basic_blocks` does not notice the `throw()` so it gives us
+    # the wrong CFG. Instead manually construct the intended CFG:
+    empty!(ir.cfg.blocks)
+    push!(ir.cfg.blocks, BasicBlock(StmtRange(1,1), [], [2,4]))
+    push!(ir.cfg.blocks, BasicBlock(StmtRange(2,2), [1], []))
+    push!(ir.cfg.blocks, BasicBlock(StmtRange(3,3), [], []))
+    push!(ir.cfg.blocks, BasicBlock(StmtRange(4,4), [1], []))
+    empty!(ir.cfg.index)
+    append!(ir.cfg.index, Int[2,3,4])
+    ir.stmts.stmt[1] = GotoIfNot(Core.Argument(2), 4)
+
+    Core.Compiler.verify_ir(ir)
+    ir = Core.Compiler.cfg_simplify!(ir)
+    @test length(ir.cfg.blocks) == 3 # should have removed block 3
```

Comment:
Verify the IR here again?

On a general note I think we have a predicate `is_terminator` or something similar and I don't think it handles this correctly

Reply:
Yeah, it doesn't - It has the same problem `compute_basic_blocks` does, which is that it expects to be able to determine a terminator from the expression alone but since #49797 it depends on the type as well

Reply:
I wonder if instead we should insert an unreachable terminator instead. Normalizing the IR more.

Reply:
I'd be in support of that, but I believe @keno is trying to avoid making any insertions in `irinterp`

Reply:
Terminator normalization is #41476, but @topolarity is right, I don't want to make irinterp structurally mutating (and thus statement-index invalidating) just for this.